Allow datagen to use vanilla armor trim textures

GizmoTheMoonPig opened the pull request:
1.20 introduced armor trims, which use dynamically generated textures for the armor overlay based on what material is used. Unfortunately, since these textures don't exist until the game is actually run, there's not an easy way to add textures like these to item models, as the ModelBuilder's `texture` method checks that the texture actually exists. 
This PR aims to fix that in an easy way, by adding a new `uncheckedTexture` method to the ModelBuilder. Using this method won't make the ModelBuilder check if the texture actually exists when generating an item model, allowing you to add things like trim overlays to your armors using datagen. 
I chose to name it `uncheckedTexture` because models use a similar naming convention with UncheckedModelFile, which doesn't check if a model actually exists when using it in datagen. If anyone has a better name proposal, please let me know and I'll change it. 
I also added to the existing `DataGeneratorTest` to show that the method works as intended. Switching `uncheckedTexture` to just use `texture` will throw an error since the texture doesn't actually exist when it's called.

GizmoTheMoonPig committed:
add uncheckedTexture to the ModelBuilder

ChampionAsh5357 commented:
Honestly, I don't think this is a bad solution since we have `UncheckedModelFile`. However, the correct solution to this would be to update `ExistingFileHelper` to consume the generated runtime resources.

GizmoTheMoonPig commented:
I'll have to look into that then. We're entering an area I'm not too familiar with so it may take me a little bit to get familiar with it. Will mark as a draft until I get this fully figured out. 

In order to switch to changing the ExistingFileHelper instead I need to make sure that this will also work with any generated resources mods make as well, not just vanilla. I'm sure there's mods out there that have their own systems to do this, or some may use vanilla's system in the future at some point.

PlatinPython commented:
Yeah, the correct solution, _imo_, would be to use `ExistingFileHelper#trackGenerated` for those textures. I'm doing that in [my mod](https://github.com/PlatinPython/RGBBlocks/blob/1.20.1/src/main/java/platinpython/rgbblocks/data/DataGatherer.java#L30-L73) and it works great.

GizmoTheMoonPig committed:
add vanilla trims to the ExistingFileHelper instead

GizmoTheMoonPig commented:
> Yeah, the correct solution, *imo*, would be to use ExistingFileHelper#trackGenerated for those textures. I'm doing that in [my mod](https://github.com/PlatinPython/RGBBlocks/blob/1.20.1/src/main/java/platinpython/rgbblocks/data/DataGatherer.java#L30-L73) and it works great.

Thank you for letting me know about this! I'm very glad I didn't have to cobble something up myself.
I've changed the PR to instead add vanilla trims to the `ExistingFileHelper` in the `ItemModelBuilder` by default, since apparently mods can just add their stuff as needed. I considered using the trim material registry for this but that would require a lot more tinkering that I don't really think is necessary, so it's using a hardcoded list of vanilla trim materials at the moment. I've removed the `uncheckedTexture` method now that this is done a different way, and the test mod has been updated to show that it works even when using the normal `texture` method again. 
I'll keep this marked as a draft just in case this isn't the way I should be going about this. Once someone from triage gives their approval, I'll unmark it.

XFactHD commented on src/main/java/net/minecraftforge/client/model/generators/ItemModelBuilder.java:
```diff
@@ -23,10 +27,18 @@
         super(outputLocation, existingFileHelper);
+        //add vanilla armor trims to the existing file helper
+        for (ResourceKey<TrimMaterial> trim : VANILLA_TRIM_MATERIALS) {
+            this.existingFileHelper.trackGenerated(new ResourceLocation("boots_trim_" + trim.location().getPath()), PackType.CLIENT_RESOURCES, ".png", "textures/trims/item");
```
From my understanding, the prefix is intended to be the general root folder of a given "resource type". In this case that would mean that the prefix should just be `textures` and the rest of your current prefix should be included in the file name, at which point it would be cleaner to use the existing `ResourceType` specified in `ModelProvider.TEXTURE` instead of specifying the equivalent pack type, prefix and suffix manually. Making said `ResourceType` field public to allow modders to use it outside of a `ModelProvider` (i.e. in a `SpriteSourceProvider` or a `BlockStateProvider`) would be a good fit for this PR as well in my opinion.

GizmoTheMoonPig replied:
I didn't technically need to make it public since ModelProvider is in the same package, but I did anyway since you mentioned that last part. Will revert that back if someone isn't happy about it

GizmoTheMoonPig committed:
use ModelProvider.TEXTURE

GizmoTheMoonPig committed:
goof

GizmoTheMoonPig committed:
directory actually uses items/, update test mod as well

GizmoTheMoonPig committed:
import fix

GizmoTheMoonPig committed:
ugh

GizmoTheMoonPig committed:
remove unused file

GizmoTheMoonPig commented:
I've updated this PR to 1.20.2, feel free to look it over whenever :)

GizmoTheMoonPig committed:
Merge branch '1.20.x' into 1.20-unchecked-texture-builder
